### PR TITLE
 Addition of new ML regression benchmarks 

### PR DESCRIPTION
**Pull request by desireevl**
### Summary
Addition of new ML benchmarks including benchmarks for regression and new dataset added.

### CODE REVIEW

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        diabetes_features = load_diabetes()["data"]
+        diabetes_labels = load_diabetes()["target"]
+
+        self.dataset_diabetes_features = diabetes_features[:35]
+        self.dataset_diabetes_labels = diabetes_labels[:35]
+
+        self.datasets = {
+            "dataset_synthetic": {
+                "features": self.dataset_synthetic_features,
+            },
```
[the remark this answers is not in the document]
Comment (desireevl) [reply machine_learning/benchmarks/base_regressor_benchmark.py]: Added new regression appropriate dataset and format is now same as those in classifier benchmarks.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        # construct simple feature map
+        param_x = Parameter("x")
+        feature_map = QuantumCircuit(1, name="fm")
+        feature_map.ry(param_x, 0)
+
+        # construct simple ansatz
+        param_y = Parameter("y")
+        ansatz = QuantumCircuit(1, name="vf")
+        ansatz.ry(param_y, 0)
```
[the remark this answers is not in the document]
Comment (desireevl) [reply machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:49]: Updated these feature maps and ansatz to be optimal for dataset

**Comment by manoelmarques** [comment]
The CI passed but the new benchmarks took 5h46min to run in asv “quick” mode. A jump from the current ~30 min. It will take much longer in “normal” mode.

**Comment by HuangJunye** [comment]
@desireevl Great work! What's the status of this PR? Can it be merged soon?

**Comment by desireevl** [comment]
I believe @adekusar-drl is still reviewing this PR

**Comment by adekusar-drl** [comment]
@dlasecki @manoelmarques Could you please review this PR?

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Base for Regressor benchmarks."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:12]:
```suggestion
"""Base class for Regressor benchmarks."""
```

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    DATASET_SYNTHETIC_REGRESSION_LABELS,
+    load_ccpp,
+)
+
+DATASET_SYNTHETIC_REGRESSION = "dataset_synthetic_regression"
+DATASET_CCPP_REGRESSION = "dataset_ccpp"
+
+
+class BaseRegressorBenchmark(ABC):
+    """Base for Regressor benchmarks."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:38]:
```suggestion
    """Base class for Regressor benchmarks."""
```

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    def __init__(self):
+
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:41]:
```suggestion
```

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        scaler = MinMaxScaler((-1, 1))
+        ccpp_features = scaler.fit_transform(ccpp_features)
+        ccpp_labels = scaler.fit_transform(ccpp_labels.reshape(-1, 1))
+
+        (
+            ccpp_train_features,
+            ccpp_test_features,
+            ccpp_train_labels,
+            ccpp_test_labels,
+        ) = train_test_split(ccpp_features, ccpp_labels, test_size=5, shuffle=False)
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:86]: I would consider making it a separate method, e.g. ```_prepare_ccpp_dataset(...)```, for improved readability.

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        num_inputs = 1
+        # construct simple feature map
+        param_x = Parameter("x")
+        feature_map = QuantumCircuit(1, name="fm")
+        feature_map.ry(param_x, 0)
+
+        # construct simple ansatz
+        param_y = Parameter("y")
+        ansatz = QuantumCircuit(1, name="vf")
+        ansatz.ry(param_y, 0)
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:115]: Code duplication could be reduced by introducing a method.

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+        opflow_qnn = TwoLayerQNN(
+            num_inputs, feature_map, ansatz, quantum_instance=self.backends[quantum_instance_name]
+        )
+
+        model = NeuralNetworkRegressor(opflow_qnn, optimizer=optimizer)
+        return model
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:122]:
```suggestion
        return NeuralNetworkRegressor(opflow_qnn, optimizer=optimizer)
```

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    ) -> NeuralNetworkRegressor:
+        num_inputs = 4
+        feature_map = ZZFeatureMap(num_inputs)
+        ansatz = RealAmplitudes(num_inputs)
+        opflow_qnn = TwoLayerQNN(
+            num_inputs, feature_map, ansatz, quantum_instance=self.backends[quantum_instance_name]
+        )
+
+        model = NeuralNetworkRegressor(opflow_qnn, optimizer=optimizer)
+        return model
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:135]:
```suggestion
        return NeuralNetworkRegressor(opflow_qnn, optimizer=optimizer)
```

File: machine_learning/benchmarks/datasets.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        [1.07553076],
+        [0.70927951],
+        [0.64405885],
+        [-0.36920025],
+        [0.89642704],
+        [0.41763132],
+        [1.13432333],
+        [-0.56352804],
+    ]
+)
```
Comment (dlasecki) [inline machine_learning/benchmarks/datasets.py:79]: A short comment on how this data was obtained would be helpful. Does it also come from the website linked below? If yes, let's mention it.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+            if dataset == DATASET_SYNTHETIC_REGRESSION:
+                model = self._construct_qnn_synthetic(
+                    quantum_instance_name=backend, optimizer=COBYLA()
+                )
+            else:
+                # we have only two dataset for now, so this clause is for "dataset_ccpp"
+                model = self._construct_qnn_ccpp(
+                    quantum_instance_name=backend, optimizer=COBYLA(maxiter=100)
+                )
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:61]: I would introduce an `elif` statement that checks for ccpp and throw an error in `else`.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self):
+        super().__init__()
+        self.model = None
+        self.train_features = None
+        self.train_labels = None
+        self.test_features = None
+        self.test_labels = None
+
+    def setup_cache(self):
+        """Cache Opflow fitted model"""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:48]:
```suggestion
        """Cache Opflow fitted model."""
```

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+    def __init__(self):
+        super().__init__()
+        self.model = None
+        self.train_features = None
+        self.train_labels = None
+        self.test_features = None
+        self.test_labels = None
+
+    def setup_cache(self):
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:47]:
```suggestion
    def setup_cache(self) -> None:
```

Comment (dlasecki) [reply machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:47]: Please check other places for the same issue.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+                    quantum_instance_name=backend, optimizer=COBYLA(maxiter=100)
+                )
+            model.fit(train_features, train_labels)
+
+            file_name = f"{dataset}_{backend}.pickle"
+            with open(file_name, "wb") as file:
+                pickle.dump(model._fit_result, file)
+
+    def setup(self, dataset: str, quantum_instance_name: str):
+        """Setup the benchmark."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:69]:
```suggestion
        """Set up the benchmark."""
```

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py (hunk 11 above)
Comment (dlasecki) [reply machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:61]: Please check other places for same issue.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        with open(file_name, "rb") as file:
+            self.model._fit_result = pickle.load(file)
+
+    # pylint: disable=invalid-name
+    def time_score_opflow_qnn_regressor(self, _, __):
+        """Time scoring OpflowQNN regressor on data."""
+        self.model.score(self.train_features, self.train_labels)
+
+    def time_predict_opflow_qnn_regressor(self, _, __):
+        """Time predicting with regressor OpflowQNN."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:92]:
```suggestion
        """Time predicting with OpflowQNN regressor."""
```

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+        self.model.predict(self.train_features)
+
+    def track_score_opflow_qnn_regressor(self, _, __):
+        """R2 score of the model on data."""
+        score = self.model.score(self.test_features, self.test_labels)
+        return score
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:98]:
```suggestion
        return self.model.score(self.test_features, self.test_labels)
```

Comment (dlasecki) [reply machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py:98]: Please check other places for the same issue.

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+
+from .base_regressor_benchmark import (
+    BaseRegressorBenchmark,
+    DATASET_SYNTHETIC_REGRESSION,
+    DATASET_CCPP_REGRESSION,
+)
+
+
+class OpflowQnnFitRegressorBenchmarks(BaseRegressorBenchmark):
+    """Opflow QNN Regressor benchmarks."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:27]:
```suggestion
    """OpflowQNN Regressor benchmarks."""
```

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+            "cobyla": COBYLA(maxiter=100),
+            "nelder-mead": NELDER_MEAD(maxiter=50),
+            "l-bfgs-b": L_BFGS_B(maxiter=20),
+        }
+        self.train_features = None
+        self.train_labels = None
+        self.model = None
+
+    def setup(self, dataset: str, quantum_instance_name: str, optimizer: str):
+        """Setup the benchmark."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:50]:
```suggestion
        """Set up the benchmark."""
```

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+        self.train_features = self.datasets[dataset]["train_features"]
+        self.train_labels = self.datasets[dataset]["train_labels"]
+
+        if dataset == DATASET_SYNTHETIC_REGRESSION:
+            self.model = self._construct_qnn_synthetic(
+                quantum_instance_name, self.optimizers[optimizer]
+            )
+        elif dataset == DATASET_CCPP_REGRESSION:
+            self.model = self._construct_qnn_ccpp(quantum_instance_name, self.optimizers[optimizer])
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:59]: Would add `else` that throws an error.

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+    # pylint: disable=invalid-name
+    def time_fit_opflow_qnn_regressor(self, _, __, ___):
+        """Time fitting OpflowQNN regressor to data."""
```
Comment (dlasecki) [inline machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:63]: Sometimes regressor is spelled with a capital letter and sometimes not. I would suggest making it consitent in all comments.

**Review by dlasecki — CHANGES_REQUESTED** [review]

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py (hunk 20 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:65]: I hope, I fixed this.

File: machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py (hunk 19 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/opflow_qnn_regressor_fit_benchmark.py:59]: Added an "else" clause.

File: machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py (hunk 11 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/opflow_qnn_regressor_benchmark.py]: Done in the regression benchmarks. Will change in the classification benchmarks as well.

File: machine_learning/benchmarks/datasets.py (hunk 10 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/datasets.py:81]: Added a short comment. The dataset is taken from the QML tutorial, this is just a noisy sine wave.

File: machine_learning/benchmarks/base_regressor_benchmark.py (hunk 7 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/base_regressor_benchmark.py:120]: I'm not quite sure it is beneficial here. Right now it is pretty clear what the feature map and ansatz are, while with a method we hide circuits in this method.

File: machine_learning/benchmarks/base_regressor_benchmark.py (hunk 6 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/base_regressor_benchmark.py]: Extracted a method that loads the dataset, pick a subset and rescale it. I keep `train_test_split` in the constructor as well as for the synthetic dataset.

**Comment by adekusar-drl** [comment]
@dlasecki could you please take a look?

File: machine_learning/benchmarks/base_regressor_benchmark.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+            },
+        }
+
+    def _prepare_ccpp(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Prepare the CCPP dataset, we can afford only a tiny subset of the dataset for training.
+        """
+        ccpp_features, ccpp_labels = load_ccpp()
+        ccpp_features = ccpp_features[:25]
+        ccpp_labels = ccpp_labels[:25]
```
Comment (dlasecki) [inline machine_learning/benchmarks/base_regressor_benchmark.py:102]:
```suggestion
        num_features = 25
        ccpp_features = ccpp_features[:num_features]
        ccpp_labels = ccpp_labels[:num_features]
```

File: machine_learning/benchmarks/datasets.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Dataset definitions for machine learning benchmarks"""
```
Comment (dlasecki) [inline machine_learning/benchmarks/datasets.py:13]:
```suggestion
"""Dataset definitions for machine learning benchmarks."""
```

**Review by dlasecki — CHANGES_REQUESTED** [review]

File: machine_learning/benchmarks/base_regressor_benchmark.py (hunk 21 above)
Comment (adekusar-drl) [reply machine_learning/benchmarks/base_regressor_benchmark.py]: replaced `num_features` to `num_samples`.

**Review by dlasecki — DISMISSED** [review]

**Review by dlasecki — APPROVED** [review]

**Pull request merged**